### PR TITLE
Rewrite equality with is null check in JOIN ON section

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Rewrite equality with `is null` check in JOIN ON section. *Analyzer only*

```sql
SET allow_experimental_analyzer = 1;
SELECT * FROM t1 JOIN t2 ON a = b OR (a IS NULL AND b IS NULL)
```
will be transformed into

```sql
SELECT * FROM t1 JOIN t2 ON a <=> b
```

It provides better compatibility with BI tools.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit b5ef6e2015ea8d0849d0012199abe840a89d5493 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56538/b5ef6e2015ea8d0849d0012199abe840a89d5493/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>❌ failure</td></tr>
<tbody>
</table>

**Comment by vdimir** [comment]
I changed changelog entry from `Performance Improvement` back to `Improvement` because such expressions in JOIN ON was not supported at all and PR adds possibility to execute such a query, so it doesn't improve a query execution speed.
And in `SELECT` expression we don't support `<=>` operator, so no change there.

File: src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp [hunk 1]
```diff
@@ -5,11 +5,211 @@
...
+
+            auto * argument_function = argument->as<FunctionNode>();
+            if (!argument_function)
+                continue;
+
+            const auto & func_name = argument_function->getFunctionName();
+            if (func_name == "equals" || func_name == "isNotDistinctFrom")
+                equals_functions_indices.push_back(or_operands.size() - 1);
+
+            if (func_name == "and")
```
Comment (antonio2368) [inline src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp:93]: `else if`

File: src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp [hunk 2]
```diff
@@ -5,11 +5,211 @@
...
+private:
+    bool need_rerun_resolve = false;
+
+    /// Returns true if type of some operand is changed and parent function needs to be re-resolved
+    static bool tryOptimizeIsNotDistinctOrIsNull(QueryTreeNodePtr & node, const ContextPtr & context)
+    {
+        auto & function_node = node->as<FunctionNode &>();
+        assert(function_node.getFunctionName() == "or");
+
+        QueryTreeNodes or_operands;
```
Comment (antonio2368) [inline src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp:64]:
```cpp
or_operands.reserve(function_node.getArguments()->getNodes().size());
```

File: src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp [hunk 3]
```diff
@@ -5,11 +5,211 @@
...
+        QueryTreeNodes new_or_operands;
+        for (size_t i = 0; i < or_operands.size(); ++i)
+        {
+            if (arguments_to_reresolve.contains(i))
+            {
+                auto * function = or_operands[i]->as<FunctionNode>();
+                if (function->getFunctionName() == "equals")
+                {
+                    /// We should replace `a = b` with `a <=> b` because we removed checks for IS NULL
+                    need_reresolve = need_reresolve || function->getResultType()->isNullable();
```
Comment (antonio2368) [inline src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp:172]:
```suggestion
                    need_reresolve ||= function->getResultType()->isNullable();
```

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**